pkgs/development/python-modules/openai-triton/default.nix

```diff
@@ -0,0 +1,199 @@
+  };
+
+  patches = [
+    # Prerequisite for llvm15 patch
+    (fetchurl {
+      url = "https://github.com/openai/triton/commit/2aba985daaa70234823ea8f1161da938477d3e02.patch";
+      hash = "sha256-HEuLZFif++a/fKs3dyIhqSc+D2DPbzEXOSSR4nRWtgQ=";
+    })
+    (fetchurl {
+      url = "https://github.com/openai/triton/commit/e3941f9d09cdd31529ba4a41018cfc0096aafea6.patch";
+      hash = "sha256-sl8woykSLCq4ZJYzEQdCPWM8rzv+s4RTK7PuqsTcmy8=";
+    })
+
+    # Not using fetchurl because needed to remove binary diff for ptxas
+    ./llvm15.patch
+  ];
+
+  postPatch = ''
+    substituteInPlace python/setup.py \
+      --replace \
+        '= get_thirdparty_packages(triton_cache_path)' \
+        '= os.environ["cmakeFlags"].split()'
+  ''
+  # A typo in setup.py?
+  # NOTE: Ok, now I'm disabling setuptools check anyway
+  # + ''
+  #   substituteInPlace python/setup.py \
+  #     --replace '"tests"' '"test"'
+  # ''
+  # Wiring triton=2.0.0 with llcmPackages_rocm.llvm=5.4.3
+  # Revisit when updating either triton or llvm
+  + ''
+    substituteInPlace CMakeLists.txt \
+      --replace "nvptx" "NVPTX" \
+      --replace "LLVM 11" "LLVM"
+    sed -i '/AddMLIR/a set(MLIR_TABLEGEN_EXE "${llvmPackages.mlir}/bin/mlir-tblgen")' CMakeLists.txt
+    sed -i '/AddMLIR/a set(MLIR_INCLUDE_DIR ''${MLIR_INCLUDE_DIRS})' CMakeLists.txt
+    find -iname '*.td' -exec \
+      sed -i \
+      -e '\|include "mlir/IR/OpBase.td"|a include "mlir/IR/AttrTypeBase.td"' \
+      -e 's|include "mlir/Dialect/StandardOps/IR/Ops.td"|include "mlir/Dialect/Func/IR/FuncOps.td"|' \
+      '{}' ';'
+    substituteInPlace unittest/CMakeLists.txt --replace "include(GoogleTest)" "find_package(GTest REQUIRED)"
+    sed -i 's/^include.*$//' unittest/CMakeLists.txt
+  '' + ''
+    sed -i '/LINK_LIBS/i NVPTXInfo' lib/Target/PTX/CMakeLists.txt
+  ''
+  + ''
+    sed -i '/LINK_LIBS/i NVPTXCodeGen' lib/Target/PTX/CMakeLists.txt
+  ''
+  # # TritonMLIRIR already links MLIRIR. Not transitive?
+  # + ''
+  #   echo "target_link_libraries(TritonPTX PUBLIC MLIRIR)" >> lib/Target/PTX/CMakeLists.txt
+  # ''
+  # Already defined in llvm, when built with -DLLVM_INSTALL_UTILS
+  + ''
+    substituteInPlace bin/CMakeLists.txt \
+      --replace "add_subdirectory(FileCheck)" ""
+
+    rm cmake/FindLLVM.cmake
+  '';
+
+  nativeBuildInputs = [
+    cmake
+    (llvmPackages.llvm.override {
+      llvmTargetsToBuild = [ "NATIVE" "NVPTX" ];
+      # Upstream CI sets these too:
+      # targetProjects = [ "mlir" ];
+      extraCMakeFlags = [
+        "-DLLVM_INSTALL_UTILS=ON"
+      ];
+    })
+    llvmPackages.mlir
+    lit
+    pythonRelaxDepsHook
+  ];
+
+  buildInputs = [
+    pybind11
+    gtest
+    zlib
+    ncurses
+    libxml2.dev
+  ];
+
+  propagatedBuildInputs = [
+    filelock
+  ];
+
+
+  # Avoid GLIBCXX mismatch with other cuda-enabled python packages
+  env.CC = "${backendStdenv.cc}/bin/cc";
+  env.CXX = "${backendStdenv.cc}/bin/c++";
```

Comment:
Hmm, as @katanallama notes in https://github.com/NixOS/nixpkgs/pull/221652, the simple `CC=` CXX=` didn't prevent `libtriton.so` linking against gcc12's libstdc++:

```
❯ patchelf  --print-rpath result/lib/python3.10/site-packages/triton/_C/libtriton.so
/nix/store/rg9b3w1rqm7b5cwr9g8d0z6jii1g462n-zlib-1.2.13/lib:/nix/store/q75826qw95k82qkmqnj3haw63zp73w2g-ncurses-6.4/lib:/nix/store/8bmp6r3a0xfha3wj36phlc47clh9w81l-glibc-2.35-224/lib:/nix/store/wvm2hvqdbbsp1f11463mrw8nyv678ipm-gcc-12.2.0-lib/lib
```

I remember that https://github.com/NixOS/nixpkgs/pull/173411 helps, but I'm forgetting now how we addressed this in torchvision&c

Reply:
UPD: apparently, variables set via `env` are not exported. Moving to `preConfigure` with `export ...=...` helps

Reply:
I've no idea how ffcv's checkPhase ran through without errors this time, but not before I used `env.CC` o_0